### PR TITLE
Warn if shadowing built-ins

### LINKED ISSUE #1249: Warn if built-in globals are shadowed
"Built-in" functions such as `sha3` and `selfdestruct` can be hidden by functions and events with the same name. This can alter the functionality of contracts in non-obvious ways.

Consider [the following contract](https://ethereum.github.io/browser-solidity/#gist=d90c5a799aab7a1595e9c62a0c80a0d4):

```
pragma solidity ^0.4.2;

contract FakingItsOwnDeath {
    function selfdestruct(address x) internal constant {}
}

contract SelfDestructive is FakingItsOwnDeath {
    function die() public {
        selfdestruct(address(0x0));
    }
}
```

At a glance, `SelfDestructive.die` appears to destroy the contract. But in fact, it's actually calling `FakingItsOwnDeath.selfdestruct` rather than the built-in version of `selfdestruct` and so the contract doesn't die.

It is also to possible to create functions/events called `this` and `now`.

(Tested using version `0.4.2+commit.af6afb04.mod.Emscripten.clang` of the online Solidity compiler.)

### PR DESCRIPTION
**Pull request by axic**
Fixes #1249.

### CODE REVIEW

File: libsolidity/analysis/StaticAnalyzer.cpp [hunk 1]
```diff
@@ -49,6 +50,11 @@ void StaticAnalyzer::endVisit(ContractDefinition const&)
 bool StaticAnalyzer::visit(FunctionDefinition const& _function)
 {
 	m_nonPayablePublic = _function.isPublic() && !_function.isPayable();
+
+	for (Declaration const* declaration: m_globals)
+		if (declaration->name() == _function.name())
+			warning(_function.location(), "Shadowing global function \"" + _function.name() + "\".");
```
Comment (axic) [inline libsolidity/analysis/StaticAnalyzer.cpp]: Perhaps it is better to say "Shadowing builtin function".

File: libsolidity/analysis/StaticAnalyzer.cpp [hunk 2]
```diff
@@ -67,6 +73,15 @@ bool StaticAnalyzer::visit(MemberAccess const& _memberAccess)
 	return true;
 }
 
+bool StaticAnalyzer::visit(VariableDeclaration const& _variable)
+{
+	for (Declaration const* declaration: m_globals)
+		if (declaration->name() == _variable.name())
+			warning(_variable.location(), "Shadowing global variable \"" + _variable.name() + "\".");
+
```
Comment (axic) [inline libsolidity/analysis/StaticAnalyzer.cpp]: Perhaps it is better to say "Shadowing builtin variable".

File: libsolidity/analysis/StaticAnalyzer.cpp [hunk 3]
```diff
@@ -49,6 +50,11 @@ void StaticAnalyzer::endVisit(ContractDefinition const&)
 bool StaticAnalyzer::visit(FunctionDefinition const& _function)
 {
 	m_nonPayablePublic = _function.isPublic() && !_function.isPayable();
+
+	for (Declaration const* declaration: m_globals)
```
Comment (axic) [inline libsolidity/analysis/StaticAnalyzer.cpp]: @chriseth any idea why here `m_globals.size() == 0`, while in the constructor it is still 18?

File: libsolidity/analysis/StaticAnalyzer.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/analysis/StaticAnalyzer.cpp]: It does not have to be a function. What about `builtin symbol`?

Comment (axic) [reply libsolidity/analysis/StaticAnalyzer.cpp]: This is within `visit(FunctionDefinition)`, but using `builtin symbol` for both cases is fine.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 4]
```diff
@@ -93,7 +93,7 @@ parseAnalyseAndReturnError(string const& _source, bool _reportWarnings = false,
 				}
 		if (success)
 		{
-			StaticAnalyzer staticAnalyzer(errors);
+			StaticAnalyzer staticAnalyzer(globalContext->declartions(), errors);
```
Comment (federicobond) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp]: declarations

Comment (axic) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp]: Thanks, there's still that other issue, at least on my computer.

File: libsolidity/analysis/StaticAnalyzer.cpp (hunk 3 above)
Comment (axic) [reply libsolidity/analysis/StaticAnalyzer.cpp]: Maybe it was only a problem on my computer, it seems to work on AppVeyor.

**Comment by axic** [comment]
Status: tests fail on Linux, but succeed on Windows.

**Comment by axic** [comment]
No idea why the original `vector<Declarations>` didn't work on linux/mac, replaced with a `vector<string>` and now including `this` and `super`. Note: we could use the `currentThis()` and `currentSuper()` but that seems like an overkill.

**Comment by chriseth** [comment]
Did you consider adding this code into `NameAndTypeResolver.cpp`? While perhaps conceptually less clean, it is much less error prone to add it there. We have the scopes, see what is visible where and so on. If we keep it as it is now and add a new declaration, we have to think about also adding it to the visitor here. As an example, you forgot to check for names of events :-)
Oh and what about `import x as msg from "y.sol";`?

**Comment by axic** [comment]
Fixed the event and import case.

> While perhaps conceptually less clean, it is much less error prone to add it there.

@chriseth Why should `StaticAnalyzer` exists in that case? :)

**Comment by chriseth** [comment]
@axic it is not about static analyzer. The main design problem is that the scopes are not persisted in the AST annotations.

**Comment by axic** [comment]
I'm not sure what is going on, for me it works on mac and linux (ubuntu), but travis fails at without any message:
```
Compiling ./scripts/../test/../std/StandardToken.sol...
```

**Comment by axic** [comment]
> The main design problem is that the scopes are not persisted in the AST annotations.

I think the scope issue which could affect this is members of a struct.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 5]
```diff
@@ -5078,6 +5078,60 @@ BOOST_AUTO_TEST_CASE(invalid_address_length)
...
+BOOST_AUTO_TEST_CASE(shadowing_builtins_ignores_struct)
+{
+	char const* text = R"(
+		contract C {
+			struct a {
+				uint msg;
+			}
+		}
+	)";
+	CHECK_WARNING(text, "Shadowing builtin symbol");
```
Comment (axic) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp]: Typo, this should be successful.

**Comment by chriseth** [comment]
I think it is better to add this change to `NameAndTypeResolver.cpp`, basically to anything that calls`DeclarationContainer::registerDeclaration()`. It would be good to perhaps add it to that function itself, but it does not have access to the error list. At that point, we could also check for any shadowing. One thing we have to check there is whether overriding a function is implemented as shadowing.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 6]
```diff
@@ -5078,6 +5078,60 @@ BOOST_AUTO_TEST_CASE(invalid_address_length)
...
+			event keccak256();
+		}
+	)";
+	CHECK_WARNING(text, "Shadowing builtin symbol");
+}
+
+BOOST_AUTO_TEST_CASE(shadowing_builtins_with_imports)
+{
+	char const* text = R"(
+	        import * as msg from "B.sol";
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp]: Perhaps also add something like `import {msg, block} as X from "B.sol"` - that should also be fine.

Comment (axic) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp]: Seems like that notation is broken currently:

```
Error: Expected "from".
               import {msg, block} as X from "B.sol"
```

Comment (axic) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp]: Parser has this comment:
```
        // import "abc" [as x];
        // import * as x from "abc";
        // import {a as b, c} from "abc";
```

`import {msg, block} from "B.sol";` works.

Comment (chriseth) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp]: Ah right. I meant `import * as X from "B.sol"`.

**Comment by axic** [comment]
Rebased for good measure.

**Comment by axic** [comment]
@roadriverrail would you be interested in looking into this?

**Comment by roadriverrail** [comment]
@axic Yes, I'll look into this soon.

**Comment by roadriverrail** [comment]
Similarly, I've gone ahead and done a rebase of this one, but I can't push to the repository.  Once we get that sorted, I can take on the comments @chriseth left.

**Comment by roadriverrail** [comment]
Just a quick ping before the weekly meeting...I had a rebase of this set up a couple weeks ago so that I could finish the work on it, but I can't push to the main solidity repo on this branch.  If someone can please help me out with this, I can get this done. @chriseth @axic

**Comment by chriseth** [comment]
@roadriverrail what is your progress on this? I would like to include this feature in the next version which can hopefully happen on monday.

**Comment by chriseth** [comment]
We cannot really test imports inside this testing framework.

**Comment by axic** [comment]
@chriseth is this ready? Needs to be rebased.

**Comment by chriseth** [comment]
Yes, this is ready.

**Comment by axic** [comment]
@chriseth you need to approve and merge

**Pull request merged**